umh-core/pkg/fsm/dataflowcomponent/reconcile.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2025 UMH Systems GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dataflowcomponent
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	internal_fsm "github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/internal/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm"
+	s6fsm "github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm/s6"
+)
+
+// Reconcile examines the DataFlowComponent and, in three steps:
+//  1. Check if a previous transition failed or if fetching external state failed; if so, verify whether the backoff has elapsed.
+//  2. Detect any external changes (e.g., a new configuration or external signals).
+//  3. Attempt the required state transition by sending the appropriate event.
+//
+// This function is intended to be called repeatedly (e.g. in a periodic control loop).
+// Over multiple calls, it converges the actual state to the desired state. Transitions
+// that fail are retried in subsequent reconcile calls after a backoff period.
+func (d *DataFlowComponent) Reconcile(ctx context.Context, tick uint64) (err error, reconciled bool) {
+	componentName := d.baseFSMInstance.GetID()
+	defer func() {
+		if err != nil {
+			d.baseFSMInstance.GetLogger().Errorf("error reconciling DataFlowComponent %s: %v", componentName, err)
+			d.PrintState()
+		}
+	}()
+
+	// Check if context is already cancelled
+	if ctx.Err() != nil {
+		return ctx.Err(), false
+	}
+
+	// Step 1: If there's a lastError, see if we've waited enough.
+	if d.baseFSMInstance.ShouldSkipReconcileBecauseOfError(tick) {
+		err := d.baseFSMInstance.GetBackoffError(tick)
+		d.baseFSMInstance.GetLogger().Debugf("Skipping reconcile for DataFlowComponent %s: %v", componentName, err)
+		return nil, false
+	}
+
+	// Step 2: Detect external changes.
+	if err := d.reconcileExternalChanges(ctx, tick); err != nil {
+		d.baseFSMInstance.SetError(err, tick)
+		d.baseFSMInstance.GetLogger().Errorf("error reconciling external changes: %s", err)
+		return nil, false
+	}
+
+	// Step 3: Attempt to reconcile the state.
+	err, reconciled = d.reconcileStateTransition(ctx)
+	if err != nil {
+		// If the instance is removed, we don't want to return an error here
+		if errors.Is(err, fsm.ErrInstanceRemoved) {
+			return nil, false
+		}
+
+		d.baseFSMInstance.SetError(err, tick)
+		d.baseFSMInstance.GetLogger().Errorf("error reconciling state: %s", err)
+		return nil, false
+	}
+
```

Comment:
in benthos, we are here reconciling the S6Manager. let's make the logic the same

Reply:
then also something like this would make sense:

```
// If either Benthos state or S6 state was reconciled, we return reconciled so that nothing happens anymore in this tick
	// nothing should happen as we might have already taken up some significant time of the avaialble time per tick, so better
	// to be on the safe side and let the rest handle in another tick
	reconciled = reconciled || s6Reconciled
```